builder/builder-next/control/control.go

```diff
@@ -256,9 +275,30 @@ func (c *Controller) Solve(ctx context.Context, req *controlapi.SolveRequest) (*
 		return nil, err
 	}
 	if req.Exporter != "" {
-		exp, err := w.Exporter(req.Exporter, c.opt.SessionManager)
-		if err != nil {
-			return nil, err
+		var exp exporter.Exporter
+		if c.opt.UseSnapshotter {
```

Comment:
The `builder/builder-next/control/control.go` implementation is 90% identical to the one in BuildKit. _However_ BuildKit v0.11 makes quite some changes in that (some of which may be breaking). Reverting some changes that are not directly relevant; the diff is only in `Controller.Solve()`; see below:

<details>

```patch
diff --git a/./builder/builder-next/control/control.go b/Users/thajeztah/go/src/github.com/moby/buildkit/control/control.go
index ba6b538d79..0d3e7976e5 100644
--- a/./builder/builder-next/control/control.go
+++ b/Users/thajeztah/go/src/github.com/moby/buildkit/control/control.go
@@ -2,7 +2,6 @@ package control
 
 import (
 	"context"
-	"strings"
 	"sync"
 	"sync/atomic"
 	"time"
@@ -43,7 +42,6 @@ type Opt struct {
 	ResolveCacheImporterFuncs map[string]remotecache.ResolveCacheImporterFunc
 	Entitlements              []string
 	TraceCollector            sdktrace.SpanExporter
-	UseSnapshotter            bool
 }
 
 type Controller struct { // TODO: ControlService
@@ -258,30 +256,9 @@ func (c *Controller) Solve(ctx context.Context, req *controlapi.SolveRequest) (*
 		return nil, err
 	}
 	if req.Exporter != "" {
-		var exp exporter.Exporter
-		if c.opt.UseSnapshotter {
-			if req.Exporter == mobyexporter.Moby {
-				req.Exporter = client.ExporterImage
-			}
-
-			exp, err = w.Exporter(req.Exporter, c.opt.SessionManager)
-			if err != nil {
-				return nil, err
-			}
-
-			if req.ExporterAttrs != nil {
-				reposAndTags, err := SanitizeRepoAndTags(strings.Split(req.ExporterAttrs["name"], ","))
-				if err != nil {
-					return nil, err
-				}
-				req.ExporterAttrs["name"] = strings.Join(reposAndTags, ",")
-				req.ExporterAttrs["unpack"] = "true"
-			}
-		} else {
-			exp, err = w.Exporter(req.Exporter, c.opt.SessionManager)
-			if err != nil {
-				return nil, err
-			}
+		exp, err := w.Exporter(req.Exporter, c.opt.SessionManager)
+		if err != nil {
+			return nil, err
 		}
 		expi, err = exp.Resolve(ctx, req.ExporterAttrs)
 		if err != nil {

```

</details>

Out of scope for this PR, but we should look if we can make changes in BuildKit to improve that.

I had a look if we could make our controller a wrapper for buildkit's controller  (forwarding all functions, except for `Solve()`), but it looks like that's difficult, as `Controller.gc()` is not exported, but looks like that needs to be triggered as part of `Solve()`. 

I haven't given it a _lot_ of thought yet, but;

- Would it work if `Controller.New()` accepted an `Exporter` as option?
- Or `Controller.Solve()` to accept an exporter as argument?
- I see part of the changes in the diff are to set `req.ExporterAttrs` (could those be properties of the exporter itself?)

Reply:
We definitely should sync with the buildkit team and see how we can move some code around to make this better 👍